casa/IO/MFFileIO.cc

```diff
@@ -32,80 +32,88 @@
 
 namespace casacore { //# NAMESPACE CASACORE - BEGIN
 
-  MFFileIO::MFFileIO (MultiFileBase& file, const String& name,
-                      ByteIO::OpenOption opt)
-    : itsFile     (file),
-      itsPosition (0),
-      itsName     (name)
+  MFFileIO::MFFileIO (const std::shared_ptr<MultiFileBase>& file,
+                      const String& name,
+                      ByteIO::OpenOption opt, Bool isNested)
+    : itsFile      (file),
+      itsPosition  (0),
+      itsName      (name),
+      itsIsNested  (isNested),
+      itsIsWritable(True)
   {
     if (opt == ByteIO::New  ||  opt == ByteIO::NewNoReplace) {
-      itsId = itsFile.fileId (name, False);
-      if (itsId >= 0) {
-        if (opt == ByteIO::NewNoReplace) {
-          throw AipsError ("MFFileIO: file " + name + " already exists");
-        }
-        itsFile.deleteFile (itsId);
-      }
-      itsId = itsFile.addFile (name);
+      itsId = itsFile->createFile (name, opt);
     } else {
-      itsId = itsFile.fileId (name);
+      itsId = itsFile->openFile (name);
+      itsIsWritable = (opt == ByteIO::Update);
     }
   }
 
   MFFileIO::~MFFileIO()
   {
-    itsFile.flush();
+    itsFile->closeFile (itsId);
   }
 
   void MFFileIO::remove()
   {
-    itsFile.deleteFile (itsId);
+    itsFile->deleteFile (itsId);
     itsId = -1;
   }
 
   Int64 MFFileIO::read (Int64 size, void* buffer, Bool throwException)
   {
-    Int64 n = itsFile.read (itsId, buffer, size, itsPosition);
+    Int64 n = itsFile->read (itsId, buffer, size, itsPosition);
     itsPosition += n;
     if (throwException  &&  n < size) {
       throw AipsError ("MFFileIO::read - incorrect number of bytes ("
 		       + String::toString(n) + " out of "
-                       + String::toString(size) + ") read for file "
-                       + itsName + " in MultiFileBase " + itsFile.fileName());
+                       + String::toString(size) + ") read for logical file "
+                       + itsName + " in MultiFileBase " + itsFile->fileName());
     }
     return n;
   }
 
   void MFFileIO::write (Int64 size, const void* buffer)
   {
-    Int64 n = itsFile.write (itsId, buffer, size, itsPosition);
+    if (!itsIsWritable) {
```

Comment:
itsFile might be writable, but the MFFileIO object is not if opened as readonly. Note that itsFile can be used by multiple MFFileIO objects.
It is the same as a device being writable, but a file on it might be not.

Reply:
Thanks for the clarification.